lib/events/dynamoevents/dynamoevents.go

```diff
@@ -890,18 +916,45 @@ func (l *Log) indexExists(tableName, indexName string) (bool, error) {
 	return false, nil
 }
 
-// createV2GSI creates the new global secondary Index and updates
+// createV2GSI creates the new global secondary index if it does not exist and updates
 // the schema to add a string key `date`.
 //
 // This does not remove the old global secondary index.
 // This must be done at a later point in time when all events have been migrated as per RFD 24.
 //
 // Invariants:
-// - The new global secondary index must not exist.
-// - This function must not be called concurrently with itself.
 // - This function must be called before the
 //   backend is considered initialized and the main Teleport process is started.
 func (l *Log) createV2GSI(ctx context.Context) error {
+	if err := backend.AcquireLock(ctx, l.backend, indexV2CreationLock, 5*time.Minute); err != nil {
+		return trace.Wrap(err)
+	}
+
+	release := func() error {
+		if err := backend.ReleaseLock(ctx, l.backend, indexV2CreationLock); err != nil {
+			return trace.Wrap(err)
+		}
+
+		return nil
+	}
+
+	v2Exists, err := l.indexExists(l.Tablename, indexTimeSearchV2)
+	if err != nil {
+		if err := release(); err != nil {
+			return trace.Wrap(err)
```

Comment:
Right now the parent error won't be propagated.

Reply:
Solved with aggregate errors from trace